### PR TITLE
wgengine/magicsock: fix DERP reader hang regression during concurrent reads

### LINKED ISSUE #1282: wgengine/magicsock: DERP reading broken from refactor?
hello has been having problems lately (#1266, etc). Its goroutines include:
```
goroutine 199 [select, 134 minutes]:
tailscale.com/wgengine/magicsock.(*Conn).runDerpReader(0xc0000d4000, 0xd3ee20, 0xc0029840c0, 0x0, 0xffff7f030328, 0xc00000c100, 0x4, 0xc0020e8180, 0xc003afe150, 0xc000182600)
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/magicsock/magicsock.go:1451 +0xb56
created by tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanOfAddr
        tailscale.com@v0.0.0-00010101000000-000000000000/wgengine/magicsock/magicsock.go:1300 +0xb05
```

Where that 134 minute select is here:

```go
                // Before we wake up ReceiveIPv4 with SetReadDeadline,                                                          
                // note that a DERP packet has arrived.  ReceiveIPv4                                                            
                // will read this field to note that its UDP read                                                               
                // error is due to us.                                                                                          
                atomic.AddInt64(&c.derpRecvCountAtomic, 1)
                // Cancel the pconn read goroutine.                                                                             
                c.pconn4.SetReadDeadline(aLongTimeAgo)

                select {
                case <-ctx.Done():
                        return
                case c.derpRecvCh <- res:
                        select {
                        case <-ctx.Done():
                                return
                        case <-didCopy:
                                continue
                        }
                }
```

The `receiveIPv4DERP` method is the only thing that reads from that channel, and it's only called:

```go
func (c *Conn) ReceiveIPv4(b []byte) (n int, ep conn.Endpoint, err error) {
        for {
                n, pAddr, err := c.pconn4.ReadFrom(b)
                if err != nil {
                        // If the pconn4 read failed, the likely reason is a DERP reader received
                        // a packet and interrupted us.
                        // It's possible for ReadFrom to return a non deadline exceeded error
                        // and for there to have also had a DERP packet arrive, but that's fine:
                        // we'll get the same error from ReadFrom later.
                        if c.derpPacketArrived() {
                                c.pconn4.SetReadDeadline(time.Time{}) // restore
                                n, ep, err = c.receiveIPv4DERP(b)
                                if err == errLoopAgain {
                                        continue
                                }
                                return n, ep, err
                        }
                        return 0, nil, err
                }
```

Where:

```go
func (c *Conn) derpPacketArrived() bool {
        rc := atomic.LoadInt64(&c.derpRecvCountAtomic)
        if rc != c.derpRecvCountLast {
                c.derpRecvCountLast = rc
                return true
        }
        return false
}
```

This was all added in the 5c619882 refactor (for performance, sigh).

But it seems that there's a bug in there.

Further, that commit was included in 1.4.0.

I have some ideas, but might not be able to look energetically enough for a few days.

/cc @danderson @crawshaw @josharian (any eyeballs welcome)

Front logo [Front conversations](https://app.frontapp.com/open/top_2rpxt)

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes #1282

### CODE REVIEW

**Review by crawshaw — APPROVED** [review]
Makes sense! Nice fix!

**Comment by bradfitz** [comment]
I found I had to flip two lines to make the tests reliable, and I don't understand why yet:

https://github.com/tailscale/tailscale/compare/e3953c212fa8ef33d346ffbdb56fd376ef7a3694..94308f8a7d9c4d9a18f399a278f8d75f26f7f534

I left a TODO there to investigate later when I get time to think.

**Comment by bradfitz** [comment]
Okay, understood now, with big comment.

No mysteries remain.

**Comment by bradfitz** [comment]
I spoke too soon.

I was running a stress test loop in the background and it just failed. Looks like a failure to wake up the UDP read:

```
=== RUN   TestDerpReceiveFromIPv4
    magicsock.go:2269: [v1] magicsock: got updated network map; 1 peers (1 with discokey)
    magicsock.go:2157: magicsock: SetPrivateKey called (init)
    magicsock.go:2694: magicsock: CreateEndpoint: key=[TksAA]: 0000000000000000000000000000000000000000000000000000000000000001.disco.tailscale:12345
    magicsock_test.go:1451: doing 250000 sends over 4 senders
panic: test timed out after 10m0s

goroutine 477 [running]:
testing.(*M).startAlarm.func1()
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1609 +0x11f
created by time.goFunc
        /home/bradfitz/sdk/go1.15/src/time/sleep.go:167 +0x52

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0001b8480, 0xc7c04d, 0x17, 0xc99448, 0x1)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1160 +0x7cf
testing.runTests.func1(0xc0001b8480)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1430 +0xa7
testing.tRunner(0xc0001b8480, 0xc000107cc8)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1108 +0x203
testing.runTests(0xc00000e520, 0x103f3e0, 0xc, 0xc, 0xc000525ad22e3f74, 0x8bb38ebf26, 0x1046320, 0x1)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1428 +0x5ab
testing.(*M).Run(0xc000198000, 0x0)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1338 +0x4ec
main.main()
        _testmain.go:69 +0x237

goroutine 494 [runnable]:
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x3, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc000402850, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1473 +0x225
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024450, 0xc000024440, 0x4, 0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x3, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 493 [runnable]:
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x2, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc000402840, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1473 +0x225
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024450, 0xc000024440, 0x4, 0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x2, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 491 [runnable]:
internal/poll.runtime_pollWait(0x7f7146acc858, 0x72, 0x0)
        /home/bradfitz/sdk/go1.15/src/runtime/netpoll.go:220 +0x55
internal/poll.(*pollDesc).wait(0xc000198318, 0x72, 0x500, 0x5dc, 0x0)
        /home/bradfitz/sdk/go1.15/src/internal/poll/fd_poll_runtime.go:87 +0xe6
internal/poll.(*pollDesc).waitRead(...)
        /home/bradfitz/sdk/go1.15/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).ReadFrom(0xc000198300, 0xc00002e600, 0x5dc, 0x5dc, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/bradfitz/sdk/go1.15/src/internal/poll/fd_unix.go:215 +0x245
net.(*netFD).readFrom(0xc000198300, 0xc00002e600, 0x5dc, 0x5dc, 0x0, 0x0, 0x0, 0xc000198300, 0x0)
        /home/bradfitz/sdk/go1.15/src/net/fd_posix.go:61 +0x7f
net.(*UDPConn).readFrom(0xc000010038, 0xc00002e600, 0x5dc, 0x5dc, 0x7f7147170008, 0xc00000e120, 0x0, 0x0)
        /home/bradfitz/sdk/go1.15/src/net/udpsock_posix.go:47 +0x93
net.(*UDPConn).ReadFrom(0xc000010038, 0xc00002e600, 0x5dc, 0x5dc, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/bradfitz/sdk/go1.15/src/net/udpsock.go:118 +0x9a
tailscale.com/wgengine/magicsock.(*RebindingUDPConn).ReadFrom(0xc00000e120, 0xc00002e600, 0x5dc, 0x5dc, 0x7b, 0xd3c5e0, 0xc0001a8000, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:2744 +0xd3
tailscale.com/wgengine/magicsock.(*Conn).ReceiveIPv4(0xc00013a000, 0xc00002e600, 0x5dc, 0x5dc, 0x7b, 0xd3c5e0, 0xc0001a8000, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1597 +0x225
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4(0xc0001b8600)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1487 +0x74e
testing.tRunner(0xc0001b8600, 0xc99448)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1108 +0x203
created by testing.(*T).Run
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1159 +0x797

goroutine 495 [runnable]:
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x4, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc000402860, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1473 +0x225
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024450, 0xc000024440, 0x4, 0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x4, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 492 [runnable]:
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x1, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc000402870, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1473 +0x225
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024450, 0xc000024440, 0x4, 0xc00013a000, 0xd3a3a0, 0xc00012a1c0, 0x1, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693
FAIL    tailscale.com/wgengine/magicsock        600.032s
FAIL
```

Ah, the deadline should only be restored when we're out of queued DERP packets. Stressing that now.

**Comment by bradfitz** [comment]
And, uh, even with the retry workaround in 6e062673e53bf6929b1eacea11cafbbc9588b7e8, I now get this:

```
=== RUN   TestDerpReceiveFromIPv4
    magicsock.go:2276: [v1] magicsock: got updated network map; 1 peers (1 with discokey)
    magicsock.go:2164: magicsock: SetPrivateKey called (init)
    magicsock.go:2701: magicsock: CreateEndpoint: key=[TksAA]: 0000000000000000000000000000000000000000000000000000000000000001.disco.tailscale:12345
    magicsock_test.go:1451: doing 250000 sends over 4 senders
panic: test timed out after 10m0s

goroutine 526 [running]:
testing.(*M).startAlarm.func1()
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1609 +0x11f
created by time.goFunc
        /home/bradfitz/sdk/go1.15/src/time/sleep.go:167 +0x52

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0001ee780, 0xc7d04d, 0x17, 0xc9a448, 0x1)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1160 +0x7cf
testing.runTests.func1(0xc0001ee780)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1430 +0xa7
testing.tRunner(0xc0001ee780, 0xc000131cc8)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1108 +0x203
testing.runTests(0xc00013c4c0, 0x1040420, 0xc, 0xc, 0xc00077dc151fdc17, 0x8bb38a55d0, 0x1047360, 0x1)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1428 +0x5ab
testing.(*M).Run(0xc0001ca000, 0x0)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1338 +0x4ec
main.main()
        _testmain.go:69 +0x237

goroutine 535 [runnable]:
sync.(*Mutex).Unlock(0xc00006a190)
        /home/bradfitz/sdk/go1.15/src/sync/mutex.go:179 +0x85
context.(*cancelCtx).Done(0xc00006a180, 0x0)
        /home/bradfitz/sdk/go1.15/src/context/context.go:366 +0x95
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x3, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc0003ae390, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1485 +0x207
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024270, 0xc000024260, 0x4, 0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x3, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 536 [runnable]:
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x4, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc000271d30, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1484 +0x3a9
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024270, 0xc000024260, 0x4, 0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x4, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 532 [select]:
tailscale.com/wgengine/magicsock.(*Conn).receiveIPv4DERP(0xc00016a580, 0xc00002e600, 0x5dc, 0x5dc, 0x7b, 0xd3d600, 0xc0004981c0, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1692 +0x1ac
tailscale.com/wgengine/magicsock.(*Conn).ReceiveIPv4(0xc00016a580, 0xc00002e600, 0x5dc, 0x5dc, 0x7b, 0xd3d600, 0xc0004981c0, 0x0, 0x0)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1636 +0xab
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4(0xc0001ee900)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1487 +0x74e
testing.tRunner(0xc0001ee900, 0xc9a448)
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1108 +0x203
created by testing.(*T).Run
        /home/bradfitz/sdk/go1.15/src/testing/testing.go:1159 +0x797

goroutine 534 [semacquire]:
sync.runtime_SemacquireMutex(0xc00006a194, 0x900000000, 0x1)
        /home/bradfitz/sdk/go1.15/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc00006a190)
        /home/bradfitz/sdk/go1.15/src/sync/mutex.go:138 +0x1c5
sync.(*Mutex).Lock(0xc00006a190)
        /home/bradfitz/sdk/go1.15/src/sync/mutex.go:81 +0x85
context.(*cancelCtx).Done(0xc00006a180, 0x0)
        /home/bradfitz/sdk/go1.15/src/context/context.go:361 +0x48
tailscale.com/wgengine/magicsock.(*Conn).sendDerpReadResult(0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x2, 0x7b, 0x4b4e, 0x0, 0x0, 0x0, 0xc0003ae3a0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1485 +0x207
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024270, 0xc000024260, 0x4, 0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x2, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1464 +0x294
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693

goroutine 533 [runnable]:
tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4.func1(0xc000024270, 0xc000024260, 0x4, 0xc00016a580, 0xd3b3c0, 0xc00006a180, 0x1, 0x4b4e, 0x0, 0x0, ...)
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1476 +0x376
created by tailscale.com/wgengine/magicsock.TestDerpReceiveFromIPv4
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock_test.go:1460 +0x693
FAIL    tailscale.com/wgengine/magicsock        600.037s
FAIL
```

That's confusing in a few ways. Will look more later.

**Comment by bradfitz** [comment]
I was holding `cmd/go` wrong. The code was fine. I was hitting `cmd/go`'s default 10 minutes `-timeout` value with `-race` and a high `-count` value.

Switched to using `http://golang.org/x/tools/cmd/stress` instead and it's fine now, even without the hacky retry attempt above.

**Review by josharian — APPROVED** [review]
Took another look through. LGTM. Stress tests here also good so far:

> 37m50s: 6075 runs so far, 0 failures

I requested a test in which we close while connections are still active, but that can happen in a separate PR.

File: wgengine/magicsock/magicsock.go [hunk 1]
```diff
@@ -1465,14 +1464,29 @@ func (c *Conn) runDerpReader(ctx context.Context, derpFakeAddr netaddr.IPPort, d
...
+	// note that a DERP packet has arrived. ReceiveIPv4
 	// will read this field to note that its UDP read
 	// error is due to us.
 	atomic.AddInt64(&c.derpRecvCountAtomic, 1)
 	// Cancel the pconn read goroutine.
 	c.pconn4.SetReadDeadline(aLongTimeAgo)
 	select {
 	case <-ctx.Done():
+		select {
+		case <-c.donec:
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply wgengine/magicsock/magicsock.go:1476]: Not explicitly, but previously we had a bunch of sends that didn't select on donec and we all sorts of deadlocks in tests shutting down, and when the server itself shut down. So now they pretty much universally include a select on donec.

The more interesting test (that I assume you were talking about) is the other case in this select: the derp reader going down without the magicsock.Conn itself going down. I'll see about adding one for that.

**Comment by bradfitz** [comment]
> I requested a test in which we close while connections are still active, but that can happen in a separate PR.

Added the test in this PR.

**Pull request merged**